lib/phoenix_live_view/tag_engine.ex

```
@@ -767,15 +803,87 @@ defmodule Phoenix.LiveView.TagEngine do
+  end
+
+  defp handle_token([], state), do: state
+
+  defp handle_macro_component(
+         [{:tag, _name, _attrs, %{macro_component: module_string} = tag_meta} | _] = tokens,
+         state
+       ) do
+    Macro.Env.required?(state.caller, Phoenix.Component) ||
+      raise ArgumentError,
+            "macro components are only supported in modules that `use Phoenix.Component`"
+
+    module = validate_module!(module_string, tag_meta, state)
+
+    {ast, rest} =
+      case Phoenix.Component.MacroComponent.build_ast(tokens) do
+        {:ok, ast, rest} -> {ast, rest}
+        {:error, message, meta} -> raise_syntax_error!(message, meta, state)
+      end
+
+    try do
+      module.transform(ast, %{file: state.file, line: tag_meta.line, env: state.caller})
+    rescue
+      e in ArgumentError ->
+        raise_syntax_error!(
+          Exception.message(e),
+          tag_meta,
+          state
+        )
+    else
+      {:ok, new_ast} ->
+        new_tokens = Phoenix.Component.MacroComponent.ast_to_tokens(new_ast)
+        continue(state, new_tokens ++ rest)
+
+      {:ok, new_ast, data} ->
+        Module.put_attribute(state.caller.module, :__macro_components__, {module, data})
+        new_tokens = Phoenix.Component.MacroComponent.ast_to_tokens(new_ast)
+        continue(state, new_tokens ++ rest)
+
+      other ->
+        raise ArgumentError,
+              "a macro component must return {:ok, ast} or {:ok, ast, data}, got: #{inspect(other)}"
+    end
+  end
+
+  defp validate_module!(module_string, tag_meta, state) do
+    module =
+      case Code.eval_string(module_string, [], state.caller) do
+        {mod, _} when is_atom(mod) ->
+          mod
+
+        _ ->
+          raise_syntax_error!(
+            "the given macro component #{inspect(module_string)} is not a valid module",
+            tag_meta,
+            state
+          )
+      end
+
+    _ = Code.ensure_compiled!(module)
+
+    if not function_exported?(module, :transform, 2) do
```

Btw, you could also simply remove this check and let it raise instead.

---

Right, but the error might be more confusing for people?

```
** (UndefinedFunctionError) function ColocatedDemoWeb.MacroTest.transform/2 is undefined or private, but the behaviour Phoenix.Component.MacroComponent expects it to be present. Did you mean:

      * transform/3

    ColocatedDemoWeb.MacroTest.transform({"div", [{"name", ".foo"}], ["\n  ", {"div", [{"foo", "barbar"}], ["Hello"], %{}}, "\n"], %{}}, %{env: #Macro.Env<aliases: [{JS, Phoenix.LiveView.JS}], context: nil, context_modules: [ColocatedDemoWeb.Demo], file: "/Users/steffen/oss/colocated_demo/lib/colocated_demo_web/demo.ex", function: {:render, 1}, functions: [{Phoenix.VerifiedRoutes, [static_integrity: 2, static_path: 2, static_url: 2, unverified_path: 3, unverified_path: 4, unverified_url: 2, unverified_url: 3]}, {ColocatedDemoWeb.CoreComponents, [back: 1, button: 1, comp_with_colocated_css: 1, comp_with_colocated_web_component: 1, comp_with_hook: 1, comp_with_runtime_hook: 1, error: 1, flash: 1, flash_group: 1, header: 1, hide: 1, hide: 2, hide_modal: 1, hide_modal: 2, icon: 1, input: 1, label: 1, list: 1, modal: 1, other_comp_with_colocated_css: 1, show: 1, show: 2, show_modal: 1, show_modal: 2, simple_form: 1, table: 1, translate_error: 1, translate_errors: 2]}, {Phoenix.HTML, [attributes_escape: 1, css_escape: 1, html_escape: 1, javascript_escape: 1, raw: 1, safe_to_string: 1]}, {Phoenix.Component, [assign: 2, assign: 3, assign_new: 3, assigns_to_attributes: 1, assigns_to_attributes: 2, async_result: 1, changed?: 2, dynamic_tag: 1, focus_wrap: 1, form: 1, inputs_for: 1, intersperse: 1, link: 1, live_component: 1, live_file_input: 1, live_flash: 2, live_img_preview: 1, live_render: 2, live_render: 3, live_title: 1, to_form: 1, to_form: 2, update: 3, upload_errors: 1, upload_errors: 2, used_input?: 1]}, {Kernel, [!=: 2, !==: 2, *: 2, **: 2, +: 1, +: 2, ++: 2, -: 1, -: 2, --: 2, /: 2, <: 2, <=: 2, ==: 2, ===: 2, =~: 2, >: 2, >=: 2, abs: 1, apply: 2, apply: 3, binary_part: 3, binary_slice: 2, binary_slice: 3, bit_size: 1, byte_size: 1, ceil: 1, div: 2, elem: 2, exit: 1, floor: 1, function_exported?: 3, get_and_update_in: 3, get_in: 2, hd: 1, inspect: 1, ...]}, {Phoenix.LiveView.Helpers, [live_patch: 1, live_patch: 2, live_redirect: 1, live_redirect: 2, live_title_tag: 1, live_title_tag: 2]}, {Phoenix.LiveView, [allow_upload: 3, attach_hook: 4, cancel_async: 2, cancel_async: 3, cancel_upload: 3, clear_flash: 1, clear_flash: 2, connected?: 1, consume_uploaded_entries: 3, consume_uploaded_entry: 3, detach_hook: 3, disallow_upload: 2, get_connect_info: 2, get_connect_params: 1, push_event: 3, push_navigate: 2, push_patch: 2, push_redirect: 2, put_flash: 3, put_private: 3, redirect: 1, redirect: 2, render_with: 2, send_update: 2, send_update: 3, send_update_after: 3, send_update_after: 4, static_changed?: 1, stream: 3, stream: 4, stream_configure: 3, stream_delete: 3, stream_delete_by_dom_id: 3, stream_insert: 3, ...]}], lexical_tracker: #PID<0.251.0>, line: 9, macro_aliases: [{JS, {{ColocatedDemoWeb.Demo, 2}, Phoenix.LiveView.JS}}], macros: [{Phoenix.VerifiedRoutes, [path: 2, path: 3, sigil_p: 2, url: 1, url: 2, url: 3]}, {Gettext.Macros, [dgettext: 2, dgettext: 3, dgettext_noop: 2, dgettext_noop_with_backend: 3, dgettext_with_backend: 3, dgettext_with_backend: 4, dngettext: 4, dngettext: 5, dngettext_noop: 3, dngettext_noop_with_backend: 4, dngettext_with_backend: 5, dngettext_with_backend: 6, dpgettext: 3, dpgettext: 4, dpgettext_noop: 3, dpgettext_noop_with_backend: 4, dpgettext_with_backend: 4, dpgettext_with_backend: 5, dpngettext: 5, dpngettext: 6, dpngettext_noop: 4, dpngettext_noop_with_backend: 5, dpngettext_with_backend: 6, dpngettext_with_backend: 7, gettext: 1, gettext: 2, gettext_comment: 1, gettext_noop: 1, gettext_noop_with_backend: 2, gettext_with_backend: 2, gettext_with_backend: 3, ngettext: 3, ngettext: 4, ngettext_noop: 2, ngettext_noop_with_backend: 3, ...]}, {Phoenix.Component.Declarative, [def: 2, defp: 2]}, {Phoenix.Component, [attr: 2, attr: 3, embed_templates: 1, embed_templates: 2, render_slot: 1, render_slot: 2, sigil_H: 2, slot: 1, slot: 2, slot: 3]}, {Kernel, [!: 1, &&: 2, ..: 0, ..: 2, ..//: 3, <>: 2, @: 1, alias!: 1, and: 2, binding: 0, binding: 1, dbg: 0, dbg: 1, dbg: 2, def: 1, defdelegate: 2, defexception: 1, defguard: 1, defguardp: 1, defimpl: 2, defimpl: 3, defmacro: 1, defmacro: 2, defmacrop: 1, defmacrop: 2, defmodule: 2, defoverridable: 1, defp: 1, defprotocol: 2, defstruct: 1, destructure: 2, get_and_update_in: 2, ...]}, {Phoenix.LiveView.Helpers, [sigil_L: 2]}, {Phoenix.LiveView, [assign_async: 3, assign_async: 4, on_mount: 1, start_async: 3, start_async: 4]}], module: ColocatedDemoWeb.Demo, requires: [Application, ColocatedDemoWeb, ColocatedDemoWeb.CoreComponents, Gettext, Gettext.Macros, Kernel, Phoenix.Component, Phoenix.Component.Declarative, Phoenix.HTML, Phoenix.LiveView, Phoenix.LiveView.Helpers, Phoenix.Template, Phoenix.VerifiedRoutes], ...>})
    (phoenix_live_view 1.1.0-dev) lib/phoenix_live_view/tag_engine.ex:850: Phoenix.LiveView.TagEngine.handle_macro_component/3
    (phoenix_live_view 1.1.0-dev) lib/phoenix_live_view/tag_engine.ex:209: Phoenix.LiveView.TagEngine.handle_body/1
    (phoenix_live_view 1.1.0-dev) expanding macro: Phoenix.Component.sigil_H/2
    lib/colocated_demo_web/demo.ex:9: ColocatedDemoWeb.Demo.render/1
```
vs
```
== Compilation error in file lib/colocated_demo_web/demo.ex ==
** (Phoenix.LiveView.Tokenizer.ParseError) lib/colocated_demo_web/demo.ex:100:5: the given macro component ColocatedDemoWeb.MacroTest does not implement the `Phoenix.LiveView.MacroComponent` behaviour
    (phoenix_live_view 1.1.0-dev) lib/phoenix_live_view/tag_engine.ex:1602: Phoenix.LiveView.TagEngine.raise_syntax_error!/3
    (phoenix_live_view 1.1.0-dev) lib/phoenix_live_view/tag_engine.ex:841: Phoenix.LiveView.TagEngine.handle_macro_component/3
    (phoenix_live_view 1.1.0-dev) lib/phoenix_live_view/tag_engine.ex:209: Phoenix.LiveView.TagEngine.handle_body/1
    (phoenix_live_view 1.1.0-dev) expanding macro: Phoenix.Component.sigil_H/2
    lib/colocated_demo_web/demo.ex:9: ColocatedDemoWeb.Demo.render/1
```

But I guess the message could be improved to just say `the given macro component ColocatedDemoWeb.MacroTest does not implement the required `transform/2` callback`